src/nvim/api/vim.c

```diff
@@ -119,12 +83,25 @@ Integer nvim_get_hl_id_by_name(String name)
   return syn_check_group(name.data, name.size);
 }
 
-Dictionary nvim__get_hl_defs(Integer ns_id, Arena *arena, Error *err)
+/// Gets highlight groups
+///
+/// @note Unlike the `:highlight` command which lists all the current highlight
+///       groups that have attributes set in human readable format, this function
+///       returns a table of all current highlight groups in a machine readable format
+///       (i.e. can be directly fed to nvim_set_hl)
+///
+/// @param ns_id Get highlight groups using namespace ns_id |nvim_get_namespaces()|.
+///              Use 0 to get global highlight groups |:highlight|.
+/// @param opts  Options dict:
+///                 - name: (string) Get a highlight definition by name.
+///                 - id: (integer) Get a highlight definition by id.
+///                 - link: (boolean, default true) Show highlight's link instead of definition.
+///
+/// @param[out] err Error details, if any.
+Dictionary nvim_get_hl(Integer ns_id, Dict(get_highlight) *val, Arena *arena, Error *err)
```

Comment:
needs (at least a minimal) `@return` doc

Reply:
`/// @return Highlight definition map or dictionary contains all highlight groups in the current namespace`

What do you think?

Reply:
Does the "or" mean the shape changes depending on the args? Any reason we shouldn't always return the same shape?

Reply:
See PR description (depends on whether a filter is specified or not). I too would prefer a single format (and it may be necessary for non-Lua customers), but that would make the deprecation harder (not just a simple change to the function called; you'd also need some postprocessing).